### DISCUSSION TITLE
Regarding detection and use of processor feature sets

### DISCUSSION THREAD

**Discussion by anzz1** [ggml-org/llama.cpp#535] [General]
I was just thinking if it would be a good idea to have the ability to detect and enable/disable various instruction sets at runtime instead of the current compile-time defines? For the windows release builds and possibly future statically linked releases for unixes (flatpaks, static docker binary builds, mac dmg's, etc.) this would also remove the need to have multiple builds for the different feature sets.

This would be easy to implement in a *very* fast and completely platform-independent way by using straight cpu instructions. For x86 it can be done simply using the cpuid `0F A2` instruction. You can retrieve all the instructions compatible with the current processor with that one simple instruction.

This is already how I check for AVX512F availability in the `windows-latest-cmake` gh action runner in here: https://github.com/ggerganov/llama.cpp/blob/34c1072e497eb92d81ee7c0e12aa6741496a41c6/.github/workflows/build.yml#L174-L181

The code part being:
```c
  unsigned int u;

  __asm{
    push ebx
    xor ecx,ecx
    mov eax, 7
    cpuid
    shr ebx, 16
    and bx, 1
    mov u, ebx
    pop ebx
  }
```

ARM I'm not too familiar with but I believe the `mrs` instruction is the one to use for it.

The only minor issue is that some compilers (looking at you, MSVC) do not support inline assembly for x64 for the reason of being too lazy to make a decent compiler. Fortunately this can be circumvented by [simply putting the instructions in an array like this](https://github.com/anzz1/gta-antispam/blob/b9bcc9f24ee9a8ad41dceb7a6726b89e2f90eb39/gta-antispam.c#L102) . That way inline assembly can be used in x64 and is also supported by all compilers.

**Comment by slaren** [comment]
I think this will need to happen at some time, but it is going to require a major refactor.

**Reply by gaby** [reply]
Is there a way to limit the feature sets when building the Docker image, so it works on more systems? Currently it only works on a selected number of systems.

**Reply by slaren** [reply]
It seems that the cmake build allows specifying the processor features, but the dockerfile uses make at the moment. Maybe this could be done by changing the dockerfiles to use cmake instead.

**Reply by anzz1** [reply]
Yeah the #defines would need to be changed but I don't think it's a major refactor really.
Luckily there already exists these functions in ggml , the defines would just need to be replaced with those

and the functions themselves
```c
int ggml_cpu_has_avx2(void) {
#if defined(__AVX2__)
    return 1;
#else
    return 0;
#endif
}
```
could be changed like this
```c
#if defined(_WIN64) || defined(__x86_64__) || defined(_M_X64)
int (__cdecl *ggml_cpu_has_avx2)(void) = (void*)"\x53\x31\xC9\xB8\x07\x00\x00\x00\x0F\xA2\xC1\xEB\x05\x48\x83\xE3\x01\x48\x89\xD8\x5B\xC3";
#elif defined(_WIN32) || defined(i386) || defined(__i386__) || defined(__i386) || defined(_M_IX86)
int (__cdecl *ggml_cpu_has_avx2)(void) = (void*)"\x53\x31\xC9\xB8\x07\x00\x00\x00\x0F\xA2\xC1\xEB\x05\x83\xE3\x01\x89\xD8\x5B\xC3";
#elif defined(__arm__)
// ... //
#endif
```

**Reply by slaren** [reply]
We should take this opportunity to split each SIMD implementation of a function into a different function, a possibly use function pointers to handle the dispatch. The functions vectorized with GGML_SIMD would also need to be split into the different architectures that GGML_SIMD supports.

**Reply by anzz1** [reply]
Yeah , definitely not call the cpuid codepath everytime a function is accessed, but rather set the pointers to functions on init. but this would inevitably add more function calls since inlining couldn't be done anymore. this should be tested how important to performance the function inlining is. that being said, there is already a PR #509 on the pipeline which does away with the inlining. if inlining isn't needed, then there's no reason really to not implement this runtime feature-dependent codepath choosing. it all hangs on determining how important inlining is to the performance. anecdotally `llama.cpp` has gotten a bit slower from the earlier days, so we'll need to be vigilant about changes like this and measure the performance to not lose the greatest aspect of this, speed.

**Reply by sw** [reply]
I agree this is going in the right direction. I don't think using byte array of machine code should be necessary, though. There's [__cpuid](https://learn.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex) on Windows for that, or is there some reason that wouldn't work?

**Reply by anzz1** [reply]
@sw its simply easier this way, as the __cpuid intrinsics are different between compilers and also i haven't checked if every and all compilers even have it. there are some cpuid libraries out there which amount to a huge mess of code trying to accommodate all the different compiler intrinsics. it just doesnt make sense to me to make such a trivial thing so complicated and wrap in big functions something that is literally a single instruction.

problem is that different compilers and platforms many times do not speak the same language which is super frustrating. but all x86 processors do speak the same language. when you speak machine code to the processor instead of trying to fight all the compilers and platforms at the same time, result is guaranteed every time.

negating the question around, i can't think of a single reason why not.

**Reply by slaren** [reply]
I don't think it should be very complicated, a version for gcc and compatible compilers and another for MSVC should do it. The advantage would be that the code would be readable and maintainable. If there happens to be a bug in that magic hex string at some point, fixing it would be a problem.

**Reply by anzz1** [reply]
The thing is that it's so simple that there can't and won't be any bugs when its done once and correctly. It's not like the x86 instruction set is suddenly going to change. However compilers do. So it's the other way around. That by using ASM it's set in stone and will never change nor introduce any bugs in the future. Also just GCC and MSVC won't cut it, Clang and MinGW have their own, the cpuid intrinsic is one of the things which like every compiler has done differently. Also anytime you are able to do something in pure assembly, you should do it. It's always the best option, but obviously not applicable in 99% of the time in the modern times. Some of them define the intrinsic as as inline-asm function anyway, but with different implementations. It makes perfect sense to run the same instructions on every platform, like it's the best thing you can get. Like if you could handcraft the assembly for the whole llama.cpp, you could achieve the very best performance, interoperability and determinism, but obviously it's nigh impossible to do that. 

Anyway I'm kinda leaning towards not implementing this, at least for now. As not making any regressions to speed is of utmost priority, the effect of un-inlining all the currently inlined funcs would need to be carefully perf'd before even thinking about this. The cost of additional function calls and failed branch predictions can be very significant particularly in the hottest functions. And those are exactly the ones being inlined and changed by doing runtime feature checks. So I kinda just realized that this idea might not be so great after all.

It's more of something to keep in mind anyway, and maybe implement in future when it's proven to not hurt performance.

**Reply by sw** [reply]
Hmm I'm conflicted about the machine code in an array. You are of course right that the x86 instruction set isn't going away. But you still have to implement a calling convention which may differ by compiler or operating system, though I don't think there's a lot of variation there on x86.

I did a double-take when I saw the base64-encoded EXE in the build script. I give you that it's clever, but it's also sneaky. It's hard to assess whether the program really does what it purports to. Transparent vs sneaky is a different dimension than benign vs malicious, and I'm not accusing you of the latter.

As for un-inlining the functions, I don't think it would be a problem. At least in the straight-forward use case of the vector dot product, which operates on quite large rows at a time.

I guess it comes down to whether we want to distribute binaries or expect people to compile it themselves. If it's the former, we will have to address this, otherwise we get bad performance on some systems *and* illegal instructions on others.

**Reply by anzz1** [reply]
Yeah the un-inlining might not be a problem at all, or it can be. I'm only saying that it should be looked into. Especially if it's a function which gets called a million times, then even the overhead of a function call matters. And there is also the thing which happens when you stack inlined functions on top of each other, that the compiler can do various optimizations to the byte code which it cannot do if the functions aren't inlined. That would however be as simple as comparing the compiled between inlining or not and see what's the case. If the functions aren't hot, it won't matter much at all to the end result if it's slightly less performant. However the hottest functions need to be assessed carefully.

You're absolutely right about the *default* x86 calling convention varying between compilers/platforms, and even in the same platform, as msvc uses __cdecl for default calling convention like everyone else but then *almost* all of the winapi functions are __stdcall, but it's only the default and can be explicitly declared as seen above (\_\_cdecl). On x64 it's ignored because all normal (non-vectorcall) functions use the same fastcall calling convention. And actually in this particular case even the x86 calling convention does not matter, because there are no arguments and the stack is not used. thus cdecl and stdcall are exactly the same. the cpuid instruction clobbers only one callee-saved register rbx/ebx which is pushed and popped to/from the stack and the value is returned in rax/eax. that's exactly the beauty of machine code that you have complete control so it will simply always work since you explicitly define in pure machine code what should happen, so it will always happen, completely independent of any external factors like platform or compiler. the cpu simply wont make 1+1=3 ever, something what a compiler can do. only thing to take care of is to make sure that the code ends up in the .text/.code page with execute rights and once thats taken care of there is simply nothing that can go wrong.

and again the different compilers mostly define their cpuid intrinsics as inline assembly anyway, but the implementation differ like how and where the values are returned exactly. simply making your own intrinsic and using it for all platforms is simpler than dealing with all the compiler variations. 

but how the cpuid is implemented is really a non-issue here anyway, the right question to ask is whether it should be done.

**Reply by slaren** [reply]
> but how the cpuid is implemented is really a non-issue here anyway, the right question to ask is whether it should be done.

What would be the alternative for distributing binaries? I am not sure that anyone downstream could effectively use llama.cpp as a library until this is fixed.

**Reply by anzz1** [reply]
> What would be the alternative for distributing binaries? I am not sure that anyone downstream could effectively use llama.cpp as a library until this is fixed.

By using git submodules? Like the example shown here in this recent issue #560 
https://github.com/Bip-Rep/sherpa/tree/main/src

Isn't the unix ecosystem designed exactly around compiling everything from source? As there are always compatibility problems with precompiled binaries on linux/unix , the way libraries vary greatly between distributions makes statically linking a non-option in most cases? Or at least I've always had trouble with linux and binaries.

I'm not arguing against you here, merely asking for clarification.

**Reply by anzz1** [reply]
Anyway I put up an example implementation of a feature like this here in case some is interested or needs such a thing:
https://github.com/anzz1/cpuid

**Reply by slaren** [reply]
>Isn't the unix ecosystem designed exactly around compiling everything from source?

I think by far the most common way to distribute software in the linux ecosystem is through pre-compiled binaries, not from source. Distributions like gentoo that insist on having the user compile everything are the exception, not the norm.

**Reply by anzz1** [reply]
> > Isn't the unix ecosystem designed exactly around compiling everything from source?
> 
> I think by far the most common way to distribute software in the linux ecosystem is through pre-compiled binaries, not from source. Distributions like gentoo that insist on having the user compile everything are the exception, not the norm.

Oh I see, shows my lack of experience in not having used desktop linuxes much at all. I have a fair amount of experience in the server space, but there it's pretty much always been (not including recent developments like docker) a choice between building from source or using the package manager where the distro maintainers handle the logistics of building the source and packaging distribution.

**Comment by kunnis** [comment]
For Windows, you can use https://learn.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?view=msvc-170  to make the cpuid calls.  (and MinGW I think can use the same includes)  By wrapping the cpuid calls, you can write one compiler specific version of cpuid, one for cpuidex.  Linux has [cpuid.h](https://github.com/torvalds/linux/blob/master/arch/x86/kvm/cpuid.h), which has wrapper functions for cpuid and cpuidex.   And since these are OS libraries, you don't have to worry about which compiler supports it.  Not sure about freeBSD, but at least you won't have to deal with MSVC.

For non x86 platforms, you already have to have a separate binary, so there's no concern with runtime processor detection (until you want to make stuff for specific arm processors, and you can let the arm devs handle that).

**Comment by kunnis** [comment]
Groups like KoboldCPP have a lot of problems with this because they mostly release a binary, so they have to kind of target the lowest common CPU instead of using detection.

I was thinking of looking into this.  I think the first step would be making a file for processor feature detection, and having someone review it.  Then we can start migrating all the guards in ggml-impl.c to use those at runtime instead of at compile time.   I was thinking that this new ``ggml-cpuid.h``  would expose a series of global bools for the various features, and an init() function to initialize all the flags based on the detected processor.  You could also set the bools after initialization to disable specific processor flags for testing or comparison without recompiling.

I'd considered other ideas, but I don't want anything that might decrease performance in performance critical code.  One thing mentioned was function pointers, but that'd end up requiring anywhere switching on platform specific code to do a function call that can't be inlined vs just a conditional jump.

@ggerganov @slaren do you have any thoughts?

**Reply by ggerganov** [reply]
AFAIU proper support requires some major changes in `ggml`. This does not seem very high-priority to me, since I believe there are alternative solutions: https://github.com/ggerganov/whisper.cpp/issues/1939#issuecomment-1986928880. How difficult is for projects such as KoboldCPP to build multiple `ggml` binaries and dynamically load the correct one? This way we delegate the feature set detection outside of `ggml` and don't have to make changes to support this. Unless there are some major obstacles in doing that in practice

There is also the following attempt: https://github.com/ggerganov/whisper.cpp/pull/1261. But I haven't looked in the details yet.

**Reply by anzz1** [reply]
For maximum performance, you'd definitely want to eliminate as much branches/conditional jumps as possible. In another words, the more static and less dynamic code you have, the faster it is. Put ggml which contains the performance-critical calculations inside a library, make the GitHub runner build this library for each possible feature-set combination, then have the frontend KoboldCPP detect the feature set on startup and have it load the correct library accordingly. This way you can have best of both worlds, a one-size-fits-all solution while also having the performance of optimizations allowed by static compilation for a specific feature-set.

**Reply by kunnis** [reply]
Ah, I should have thought of compiling the library multiple times with different settings.  Good call.
